File: Algo/python/Combinations.py
''' https://leetcode.com/problems/combinations/ '''

class Solution(object):
    def __init__(self):
        self.res = []
    
    def helper(self, cur_set, k, low, high):
        if len(cur_set) == k:
            self.res.append(cur_set)
            return
        else:
            for i in range(low, high + 1):
                self.helper(cur_set + [i], k, i + 1, high)
                
    def combine(self, n, k):
        """
        :type n: int
        :type k: int
        :rtype: List[List[int]]
        """
        if k == 0:
            return self.res
        self.helper([], k, 1, n)
        return self.res
        

File: Algo/Java/Combinations.java
/* https://leetcode.com/problems/combinations/ */
import java.util.ArrayList;
import java.util.List;

class Combinations
{
	public List<List<Integer>> combine(int n, int k) 
	{
   		List<List<Integer>> list = new ArrayList<>();
   		helper(n, k, 1, list, new ArrayList<>());
   		//System.out.println(list);
   		return list;
    }

    public void helper(int n, int k, int start, List<List<Integer>> list, List<Integer> l)
    {
    	if(l.size()==k)
    	{
    		//System.out.println(l);
    		list.add(new ArrayList<>(l));
    		//System.out.println(list);
    		return;
    	}
    	for(int i=start;i<=n;i++)
    	{
    		if(!l.contains(i))
    		{
    			l.add(i);
    			helper(n, k, i, list, l);
    			l.remove(l.size()-1);
    		}
    	}
    }

	public static void main(String args[])
	{
		Combinations c = new Combinations();
		c.combine(1, 1);
	}
}